tools/jdbc/src/jni/duckdb_java.cpp

```diff
@@ -866,3 +879,91 @@ JNIEXPORT void JNICALL Java_org_duckdb_DuckDBNative_duckdb_1jdbc_1appender_1appe
+	        TableFilterSet *filters = nullptr);
+
+	static void GetSchema(uintptr_t stream_factory_ptr, ArrowSchemaWrapper &schema);
+
+private:
+	JNIEnv *env;
+	jobject array_stream_wrapper_j;
+};
+
+unique_ptr<ArrowArrayStreamWrapper> JavaArrowStreamFactory::Produce(
+    uintptr_t stream_factory_ptr,
+    std::pair<std::unordered_map<idx_t, string>, std::vector<string>> &project_columns, // unsupported
+    TableFilterSet *filters) {                                                          // unsupported
+
+	JavaArrowStreamFactory *stream_factory = (JavaArrowStreamFactory *)stream_factory_ptr;
+	auto env = stream_factory->env;
+	auto array_stream_wrapper_j = stream_factory->array_stream_wrapper_j;
+	auto wrapper = make_unique<ArrowArrayStreamWrapper>();
+
+	env->CallVoidMethod(array_stream_wrapper_j, J_DuckArrayStreamWrapper_exportArrayStream,
+	                    (long)&wrapper->arrow_array_stream);
+	if (env->ExceptionCheck()) {
+		env->Throw(env->ExceptionOccurred());
+		// shouldn't get here?
+		return nullptr;
+	}
+
+	return wrapper;
+}
+
+void JavaArrowStreamFactory::GetSchema(uintptr_t stream_factory_ptr, ArrowSchemaWrapper &schema) {
+	JavaArrowStreamFactory *stream_factory = (JavaArrowStreamFactory *)stream_factory_ptr;
+
+	auto env = stream_factory->env;
+	auto array_stream_wrapper_j = stream_factory->array_stream_wrapper_j;
+
+	env->CallVoidMethod(array_stream_wrapper_j, J_DuckArrayStreamWrapper_exportSchema, (long)&(schema.arrow_schema));
+	if (env->ExceptionCheck()) {
+		env->Throw(env->ExceptionOccurred());
+
+		return;
+	}
+
+	return;
+}
+
+JNIEXPORT void JNICALL Java_org_duckdb_DuckDBNative_duckdb_1jdbc_1ingest_1arrow_1table(JNIEnv *env, jclass,
+                                                                                       jobject conn_ref_buf,
+                                                                                       jbyteArray schema_name_j,
+                                                                                       jbyteArray table_name_j,
+                                                                                       jobject array_stream_wrapper_j) {
+	auto conn_ref = (Connection *)env->GetDirectBufferAddress(conn_ref_buf);
+	if (!conn_ref || !conn_ref->context) {
+		env->ThrowNew(J_SQLException, "Invalid connection");
+		return;
+	}
+	auto schema_name = byte_array_to_string(env, schema_name_j);
+	auto table_name = byte_array_to_string(env, table_name_j);
+
+	try {
+		auto stream_factory = make_unique<JavaArrowStreamFactory>(env, array_stream_wrapper_j);
+
+		vector<Value> params;
+		params.push_back(duckdb::Value::POINTER(reinterpret_cast<uintptr_t>(stream_factory.get())));
+		params.push_back(duckdb::Value::POINTER((uintptr_t)JavaArrowStreamFactory::Produce));
+		params.push_back(duckdb::Value::POINTER((uintptr_t)JavaArrowStreamFactory::GetSchema));
+		// TODO batch size?
+		params.push_back(duckdb::Value::UBIGINT(100000));
```

Comment:
Note to self, if https://github.com/duckdb/duckdb/pull/4539 goes in, I'll need to remove this. see https://github.com/duckdb/duckdb/pull/4539/files#diff-7c2052a992507edd04315acf93ff1c68628dd402ddd4c0658db600d9af5674aaL578